runtime/src/iree-amd-aie/driver/xrt/direct_command_buffer.cc

```diff
@@ -0,0 +1,446 @@
+  IREE_ASSERT_ARGUMENT(device);
+  IREE_ASSERT_ARGUMENT(out_command_buffer);
+  *out_command_buffer = NULL;
+  if (binding_capacity > 0) {
+    // TODO(#10144): support indirect command buffers with binding tables.
+    return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                            "indirect command buffers not yet implemented");
+  }
+
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_direct_command_buffer_t* command_buffer = NULL;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_allocator_malloc(host_allocator, sizeof(*command_buffer),
+                                (void**)&command_buffer));
+
+  iree_hal_command_buffer_initialize(
+      device, mode, command_categories, IREE_HAL_QUEUE_AFFINITY_ANY,
+      binding_capacity, &iree_hal_xrt_direct_command_buffer_vtable,
+      &command_buffer->base);
+  command_buffer->host_allocator = host_allocator;
+
+  *out_command_buffer = &command_buffer->base;
+  IREE_TRACE_ZONE_END(z0);
+  iree_arena_initialize(block_pool, &command_buffer->arena);
+  return iree_hal_resource_set_allocate(block_pool,
+                                        &command_buffer->resource_set);
+}
+static void iree_hal_xrt_direct_command_buffer_destroy(
+    iree_hal_command_buffer_t* base_command_buffer) {
+  iree_hal_xrt_direct_command_buffer_t* command_buffer =
+      iree_hal_xrt_direct_command_buffer_cast(base_command_buffer);
+  iree_allocator_t host_allocator = command_buffer->host_allocator;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  for (iree_host_size_t i = 0; i < command_buffer->set_count; ++i) {
+    // TODO: cache this information in the kernel info to avoid recomputation.
+    for (iree_host_size_t j = 0;
+         j < command_buffer->descriptor_sets[i].binding_count; ++j) {
+           // these are sub buffers allocated on the heap and hence need to be deleted
+           try {
+             delete command_buffer->descriptor_sets[i].bindings[j];
+           }
+           catch(...) {
+            iree_status_from_code(IREE_STATUS_DATA_LOSS);
+           }
+    }
+  }
+  iree_hal_resource_set_free(command_buffer->resource_set);
+  iree_arena_deinitialize(&command_buffer->arena);
+  iree_allocator_free(host_allocator, command_buffer);
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+bool iree_hal_xrt_direct_command_buffer_isa(
+    iree_hal_command_buffer_t* command_buffer) {
+  return iree_hal_resource_is(&command_buffer->resource,
+                              &iree_hal_xrt_direct_command_buffer_vtable);
+}
+
+static iree_status_t iree_hal_xrt_direct_command_buffer_begin(
+    iree_hal_command_buffer_t* base_command_buffer) {
+  // nothing to do
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_xrt_direct_command_buffer_end(
+    iree_hal_command_buffer_t* base_command_buffer) {
+  iree_hal_xrt_direct_command_buffer_t* command_buffer =
+      iree_hal_xrt_direct_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_arena_reset(&command_buffer->arena);
```

Comment:
fyi this snuck in from cuda2
https://github.com/openxla/iree/blob/main/experimental/cuda2/stream_command_buffer.c#L190
where it must be wrong as well given that we dont have inline execution there, (I think must have just been copied from cuda HAL)

Reply:
note that the above comment is in the context of the existing impl, where we directly issue the commands when _recording_ from the perspective of the whole program. If with deferred command buffer, when we reach this, we are actually not recording anymore; it's _execution_, from the perspective of the whole program. Because we already recorded into the derferred command buffer. From the perspective of the xrt hal itself, it still looks like we are recording here. (Yeah it's a bit confusing here; let me know if you want to jump into a call to chat directly.) So this way works with deferred command buffer (like how cuda2 is done) but not without it (which is the current impl)>

Reply:
oh so since I am implementing deferred command buffer for XRT, I should Ieave this in?

Reply:
Yes.